### DISCUSSION TITLE
Member is unkown

### DISCUSSION THREAD

**Discussion by robertsawko** [microsoft/pyright#4901] [Q&A] [answered]
Hello,

I am using `pyright` in neovim through an lsp plugin. It worked mostly out of the box, but there are some functions it is not finding from standard packages. This may be the way I am using them or still something in my configuration.

Minimal example:
```python
from matplotlib.pyplot import figure

fig = figure()
fig.gca().plot([0, 1, 2])
fig.savefig("a.png")
```
The code above runs fine, but pyright highlights an error:
```
Pyright: Cannot access member "savefig" for type "FigureBase"
Member "savefig" is unkown
```
Can you please advise?

**Comment by erictraut** [comment, marked as answer]
Matplotlib is not a typed library, so by default pyright will treat all symbols imported from this library as `Unknown`. Do you have type stubs installed for the library? Or perhaps you've set the `useLibraryCodeForTypes` setting to true? If so, pyright will attempt to infer types from the Python code in the library.

Inferring types from an untyped library is not foolproof, and the resulting types will not always be correct. They're usually sufficient for informing completion suggestions, but they're often not sufficient for accurate type checking. If you want to use pyright for type checking, it's important to have high-quality type information for the libraries you use. If matplotlib is important to you, I recommend contacting the maintainers and requesting that they add inlined type information. Most popular Python libraries have added such type information over the past few years, but matplotlib unfortunately hasn't yet.

In the meantime, you may want to use type stubs to provide the missing type information. Pylance ships with partial stubs for matplotlib. You can find them in [this repo](https://github.com/microsoft/python-type-stubs). Create a `typings` directory within your project and then copy the `maplotlib` directory into it. Pylance will then use these stubs rather than attempting to infer types from the library source code.

**Reply by robertsawko** [reply]
Can I just check, @erictraut. Does this repo works with Pylance only or pyright as well? I tried to create `typings` directory but that didn't do anything.

I also tried `useLibraryCodeForTypes`, by creating `pyrightconfig.json` and adding:

```json
{
  useLibraryCodeForTypes: "true"
}
```
and that seems to be working! :tada:
 
I would be happy to support to add type information into important packages like matplotlib. [This](https://github.com/matplotlib/matplotlib/issues/25590) conversation seems relevant, but I am not certain

**Reply by erictraut** [reply]
Type stubs are part of the Python typing standard, so they should work with any type checker, including pyright and mypy.

Good news — I just learned that the maintainers of matplot lib just recently incorporated stub files into the library! That means the next release should contain this type information, and separate type stubs will no longer be required! This was merged just last week as part of [this PR](https://github.com/matplotlib/matplotlib/commit/9b341f1fcc1a61c072dc643ba570134c146b20f3).
